protocols/pnet/src/crypt_writer.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::{
+    io::{self, AsyncWrite},
+    ready,
+    task::{Context, Poll},
+};
+use log::trace;
+use pin_project::pin_project;
+use salsa20::{stream_cipher::SyncStreamCipher, XSalsa20};
+use std::{fmt, pin::Pin};
+
+/// A writer that encrypts and forwards to an inner writer
+#[pin_project]
+pub struct CryptWriter<W> {
+    #[pin]
+    inner: W,
+    buf: Vec<u8>,
+    cipher: XSalsa20,
+}
+
+impl<W: AsyncWrite> CryptWriter<W> {
+    /// Creates a new `CryptWriter` with the specified buffer capacity.
+    pub fn with_capacity(capacity: usize, inner: W, cipher: XSalsa20) -> CryptWriter<W> {
+        CryptWriter {
+            inner,
+            buf: Vec::with_capacity(capacity),
+            cipher,
+        }
+    }
+
+    /// Gets a pinned mutable reference to the inner writer.
+    ///
+    /// It is inadvisable to directly write to the inner writer.
+    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut W> {
+        self.project().inner
+    }
+}
+
+/// Write the contents of a Vec<u8> into an AsyncWrite.
+///
+/// The handling 0 byte progress and the Interrupted error was taken from BufWriter in async_std.
+///
+/// If this fn returns Ready(Ok(())), the buffer will be completely flushed.
```

Comment:
That isn't true anymore.

Reply:
Why not?

- If the while loop completes normally, everything up to `len` has been written, the buffer is drained, and therefore empty / completely flushed.

- All break cases are either for `Poll::Ready(Err(whatever))` or for `Poll::Pending`. In these cases the buffer might not be empty, but the part that has been written is nevertheless drained.

I guess the sentence should read "If this fn returns Ready(Ok(())), the buffer has been completely flushed and therefore is empty", otherwise it is a bit ambiguous.

Reply:
Never mind. As my colleague pointed out, I missed setting ret to Pending.

Reply:
I still have not found a way to write these poll loops that is not somewhat jarring... Tail recursion would be so much nicer...